force_bdss/mco/optimizers/scipy_optimizer.py

```diff
@@ -0,0 +1,265 @@
+    pass
+
+
+@provides(IOptimizer)
+class ScipyOptimizer(HasStrictTraits):
+    """ Optimization of an objective function using scipy.
+    """
+
+    #: Algorithms available to work with
+    algorithms = Enum("SLSQP", "Nelder-Mead", "Powell", "CG", "BFGS",
+                      "Newton-CG", "L-BFGS-B", "TNC", "COBYLA",
+                      "trust-constr", "dogleg",
+                      "trust-ncg", "trust-exact", "trust-krylov")
+
+    def optimize_function(self, func, params):
+        """ Minimize the passed function.
+
+        Parameters
+        ----------
+        func: Callable
+            The MCO function to optimize
+            Takes a list of MCO parameter values.
+        params: list of MCOParameter
+            The MCO parameter objects corresponding to the parameter values.
+
+        Yields
+        ------
+        list of float or list:
+            The list of parameter values.
+            A float if the parameter is a RangedMCO type.
+            A list if the parameter is a RangedVector type.
+
+        Exception
+        ---------
+        ScipyTypeError
+            If params has no RangedMCO or RangedVector.
+        """
+        # verify that all parameters are Ranged or RangedVector
+        # (see the notes for this method)
+        self.verify_mco_parameters(params)
+
+        # create a "translated" function that only takes a single
+        # numpy array as the parameter argument.
+        tfunc = partial(self.translated_function, func=func, params=params)
+
+        # get the intitial parameter values and their bounds.
+        x0, bounds = self.get_initial_and_bounds(params)
+
+        # optimize the function
+        optimization_result = scipy_optimize.minimize(
+            tfunc,
+            x0,
+            method=self.algorithms,
+            bounds=bounds
+        )
+
+        # get the optimal point (list of optimal parameter values)
+        optimal_point = self.translate_array_to_mco(
+            optimization_result.x, params)
+        yield optimal_point
+
+    def translated_function(self, array, func, params):
+        """ A wrapper around the MCO function, where the
+        MCO parameter list is replaced by a numpy array.
+
+        Parameters
+        ----------
+        array: numpy.array
+            The numpy array.
+        func: Callable
+            The MCO function that takes a list of MCO parameter values.
+        params: list of MCOParameter
+            The MCO parameter objects corresponding to the parameter values.
+
+        Return
+        ------
+        array: numpy.array
+            The optimized
+        """
+
+        # Translate the numpy array into an MCO parameter list
+        param_values = self.translate_array_to_mco(array, params)
+
+        # Call the function that takes a list of MCO parameter values
+        objective = func(param_values)
+
+        # Translate the objective (kpis) into a scalar if it is not.
+        # ??????????
```

Comment:
Was this unfinished? Would something like:

```python
if not np.isscalar(objective):
    objective = np.sum(objective)
```

be suitable (if this behaviour is documented)?